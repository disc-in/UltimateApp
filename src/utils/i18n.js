import { I18n } from 'i18n-js';
import * as Localization from 'expo-localization';

import en from './locales/en';
import fr from './locales/fr';

const i18n = new I18n({
  en,
  fr,
});

i18n.defaultLocale = 'en';
i18n.locale = Localization.locale;
i18n.fallbacks = true;
i18n.enableFallback = true;

export default i18n;
