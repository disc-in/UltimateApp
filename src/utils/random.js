import * as Crypto from 'expo-crypto';

export const generateRandomHex = (size = 10) =>
  [...Array(size)].map(() => Math.floor(Math.random() * 16).toString(16)).join('');

export const generateUuid = () => Crypto.randomUUID();
