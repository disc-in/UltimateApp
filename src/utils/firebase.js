import { Platform, InteractionManager } from 'react-native';
import Constants from 'expo-constants';
import { initializeApp, setLogLevel } from 'firebase/app';
import { getDatabase, ref, set, child, get } from 'firebase/database';

import { generateRandomHex } from './random';

// Work around issue `Setting a timer for long time`
// see: https://github.com/firebase/firebase-js-sdk/issues/97
const _setTimeout = global.setTimeout;
const _clearTimeout = global.clearTimeout;
const MAX_TIMER_DURATION_MS = 60 * 1000;
if (Platform.OS === 'android') {
  const timerFix = {};
  const runTask = (id, fn, ttl, args) => {
    const waitingTime = ttl - Date.now();
    if (waitingTime <= 1) {
      InteractionManager.runAfterInteractions(() => {
        if (!timerFix[id]) {
          return;
        }
        delete timerFix[id];
        fn(...args);
      });
      return;
    }

    const afterTime = Math.min(waitingTime, MAX_TIMER_DURATION_MS);
    timerFix[id] = _setTimeout(() => runTask(id, fn, ttl, args), afterTime);
  };

  global.setTimeout = (fn, time, ...args) => {
    if (MAX_TIMER_DURATION_MS < time) {
      const ttl = Date.now() + time;
      const id = '_lt_' + Object.keys(timerFix).length;
      runTask(id, fn, ttl, args);
      return id;
    }
    return _setTimeout(fn, time, ...args);
  };

  global.clearTimeout = (id) => {
    if (typeof id === 'string' && id.startsWith('_lt_')) {
      _clearTimeout(timerFix[id]);
      delete timerFix[id];
      return;
    }
    _clearTimeout(id);
  };
}
// End of workaround

const firebaseConfig = {
  databaseURL: Constants.expoConfig.extra.firebaseDatabaseUrl,
};

const app = initializeApp(firebaseConfig);
setLogLevel('silent');

const reference = (namespace, identifier) => {
  return ref(getDatabase(app), `${namespace}/${identifier}`);
};

export const upload = async (namespace, record) => {
  const withoutUndefineds = JSON.parse(JSON.stringify(record));
  const identifier = generateRandomHex();
  await set(reference(namespace, identifier), withoutUndefineds);
  return identifier;
};

export const download = (namespace, identifier) => {
  const dbRef = ref(getDatabase(app));
  return get(child(dbRef, `${namespace}/${identifier}`))
    .then((snapshot) => {
      if (snapshot.exists()) {
        return snapshot.val();
      } else {
        return null;
      }
    })
    .catch((error) => {
      console.error(error);
    });
};
