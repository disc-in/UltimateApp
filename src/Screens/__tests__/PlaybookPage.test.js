import React from 'react';
import { connect, Provider } from 'react-redux';
import { render, fireEvent } from '@testing-library/react-native';

import store from '../../Store/testStore';
import Drill from '../../Components/animation/Drill';
import * as random from '../../utils/random';

import ConnectedPlaybookPage, { PlaybookPage } from '../PlaybookPage';

describe('<PlaybookPage />', () => {
  const play = {
    uuid: undefined,
    animation: new Drill(),
    title: 'title',
  };

  it('renders correctly when empty', () => {
    const navigation = { navigate: jest.fn(), setOptions: jest.fn() };

    const { toJSON } = render(
      <Provider store={store}>
        <ConnectedPlaybookPage navigation={navigation} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with plays', () => {
    const navigation = { navigate: jest.fn(), setOptions: jest.fn() };
    const MockedConnectedPlaybookPage = connect(() => ({ customPlays: [play] }))(PlaybookPage);

    const { toJSON } = render(
      <Provider store={store}>
        <MockedConnectedPlaybookPage navigation={navigation} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('links to drill page', async () => {
    const navigation = { navigate: jest.fn(), setOptions: jest.fn() };

    const MockedConnectedPlaybookPage = connect(() => ({ customPlays: [play] }))(PlaybookPage);
    const { getByText } = render(
      <Provider store={store}>
        <MockedConnectedPlaybookPage navigation={navigation} />
      </Provider>,
    );
    await fireEvent.press(getByText(play.title));

    expect(navigation.navigate).toBeCalledWith('PlayEditorPage', { currentPlay: play });
  });

  it('creates a new play and links to drill page', async () => {
    const navigation = { navigate: jest.fn(), setOptions: jest.fn() };
    const savePlay = jest.fn();
    jest.spyOn(random, 'generateUuid').mockImplementation(() => '123456');

    const MockedConnectedPlaybookPage = connect(
      () => ({ customPlays: [play] }),
      () => ({ savePlay }),
    )(PlaybookPage);
    const { getByTestId } = render(
      <Provider store={store}>
        <MockedConnectedPlaybookPage navigation={navigation} savePlay={savePlay} />
      </Provider>,
    );
    await fireEvent.press(getByTestId('createPlay'));

    const expectedPlay = {
      animation: new Drill(),
      title: 'Unnamed play',
      uuid: '123456',
    };
    expect(navigation.navigate).toBeCalledWith('PlayEditorPage', { currentPlay: expectedPlay });
    expect(savePlay).toBeCalledWith(expectedPlay);
  });
});
