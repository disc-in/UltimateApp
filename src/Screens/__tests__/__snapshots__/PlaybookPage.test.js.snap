// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaybookPage /> renders correctly when empty 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "marginTop": "30%",
        "padding": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 16,
          "marginBottom": 50,
          "textAlign": "center",
        }
      }
    >
      Your playbook is empty!
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderTopColor": "#e5e5e5",
        "borderTopWidth": 1,
        "opacity": 1,
        "paddingBottom": 20,
        "paddingTop": 10,
      }
    }
    testID="createPlay"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#2a9e91",
          "borderRadius": 25,
          "borderWidth": 1,
          "height": 50,
          "justifyContent": "center",
          "width": 50,
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;

exports[`<PlaybookPage /> renders correctly with plays 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
              "flexDirection": "row",
            },
            Object {
              "height": 60,
              "paddingLeft": 20,
            },
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flex": 1,
              "flexBasis": "100%",
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            title
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "opacity": 1,
              "padding": 10,
            }
          }
          testID="edit"
        >
          <Text />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "opacity": 1,
              "padding": 10,
            }
          }
          testID="delete"
        >
          <Text />
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderTopColor": "#e5e5e5",
        "borderTopWidth": 1,
        "opacity": 1,
        "paddingBottom": 20,
        "paddingTop": 10,
      }
    }
    testID="createPlay"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#2a9e91",
          "borderRadius": 25,
          "borderWidth": 1,
          "height": 50,
          "justifyContent": "center",
          "width": 50,
        }
      }
    >
      <Text />
    </View>
  </View>
</View>
`;
